Refactor WriteableIngestDocument

joegallo opened the pull request:
Rather than passing a copy of the `IngestDocument` to `SimulateProcessorResult`, we can have `WriteableIngestDocument` make a copy internally -- that makes the correct usage the only/automatic way of doing things. Also adds some error checking to `IngestDocument`'s copy constructor.

joegallo committed:
Add self-reference checking to IngestDocument's copy constructor

joegallo committed:
Internalize copying inside WriteableIngestDocument

rather than having callers copy the IngestDocument prior to calling it

joegallo committed:
Handle the failure case for WriteableIngestDocument

joegallo committed:
Add some tests

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

masseyke commented on server/src/main/java/org/elasticsearch/action/ingest/WriteableIngestDocument.java:
```diff
@@ -83,9 +83,14 @@ final class WriteableIngestDocument implements Writeable, ToXContentFragment {
     WriteableIngestDocument(IngestDocument ingestDocument) {
         assert ingestDocument != null;
-        this.ingestDocument = ingestDocument;
+        this.ingestDocument = new IngestDocument(ingestDocument); // internal defensive copy
```
I'm not concerned about simulate performance, but this does mean we'll make an extra deepCopy and do an extra reference check every time the xcontent parser runs (line 60 above). I don't think that's performance-sensitive though right?

masseyke replied:
Joe pointed out offline that WriteableIngestDocument is only used for the simulate API, so my performance concern here is irrelevant.

joegallo committed:
Add a private helper constructor

and use it to avoid unnecessarily copying the IngestDocument when it's
being built directly from the passed-in maps and cannot be
self-referencing.

joegallo committed:
Merge branch 'main' into refactor-writeable-ingest-document

masseyke reviewed (approved):
LGTM
